core/src/main/java/org/infinispan/util/concurrent/locks/impl/InfinispanLock.java

```diff
@@ -0,0 +1,555 @@
+      this.releaseRunnable = releaseRunnable;
+   }
+
+   /**
+    * Tests purpose only!
+    */
+   public void setTimeService(TimeService timeService) {
+      if (timeService != null) {
+         this.timeService = timeService;
+      }
+   }
+
+   /**
+    * It tries to acquit this lock.
+    * <p>
+    * If it is invoked multiple times with the same owner, the same {@link ExtendedLockPromise} is returned until it has
+    * timed-out or {@link #release(Object)}  is invoked.
+    * <p>
+    * If the lock is free, it is immediately acquired, otherwise the lock owner is queued.
+    *
+    * @param lockOwner the lock owner who needs to acquire the lock.
+    * @param time      the timeout value.
+    * @param timeUnit  the timeout unit.
+    * @return an {@link ExtendedLockPromise}.
+    * @throws NullPointerException if {@code lockOwner} or {@code timeUnit} is {@code null}.
+    */
+   public ExtendedLockPromise acquire(Object lockOwner, long time, TimeUnit timeUnit) {
+      Objects.requireNonNull(lockOwner, "Lock Owner should be non-null");
+      Objects.requireNonNull(timeUnit, "Time Unit should be non-null");
+
+      if (trace) {
+         log.tracef("Acquire lock for %s. Timeout=%s (%s)", lockOwner, time, timeUnit);
+      }
+
+      LockPlaceHolder lockPlaceHolder = lockOwners.get(lockOwner);
+      if (lockPlaceHolder != null) {
+         if (trace) {
+            log.tracef("Lock owner already exists: %s", lockPlaceHolder);
+         }
+         return lockPlaceHolder;
+      }
+
+      lockPlaceHolder = createLockInfo(lockOwner, time, timeUnit);
+      LockPlaceHolder other = lockOwners.putIfAbsent(lockOwner, lockPlaceHolder);
+
+      if (other != null) {
+         if (trace) {
+            log.tracef("Lock owner already exists: %s", other);
+         }
+         return other;
+      }
+
+      if (trace) {
+         log.tracef("Created a new one: %s", lockPlaceHolder);
+      }
+
+      pendingRequest.add(lockPlaceHolder);
+      tryAcquire();
+      return lockPlaceHolder;
+   }
+
+   /**
+    * It tries to release the lock held by {@code lockOwner}.
+    * <p>
+    * If the lock is not acquired (is waiting or timed out/deadlocked) by {@code lockOwner}, its {@link
+    * ExtendedLockPromise} is canceled. If {@code lockOwner} is the current lock owner, the lock is released and the
+    * next lock owner available will acquire the lock. If the {@code lockOwner} never tried to acquire the lock, this
+    * method does nothing.
+    *
+    * @param lockOwner the lock owner who wants to release the lock.
+    * @throws NullPointerException if {@code lockOwner} is {@code null}.
+    */
+   public void release(Object lockOwner) {
+      Objects.requireNonNull(lockOwner, "Lock Owner should be non-null");
+
+      if (trace) {
+         log.tracef("Release lock for %s.", lockOwner);
+      }
+
+      LockPlaceHolder wantToRelease = lockOwners.get(lockOwner);
```

Comment:
I was thinking you could save the lock promise in the command/context, and then you wouldn't need to lock again. But I guess I'd better shut up and work on the sequential interceptors instead, because that would help :)